packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
 import { CodeMaker } from 'codemaker';
-import { Assembly, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { Assembly, Type, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { join } from 'path';
 import { GoClass, Enum, Interface } from './types';
 
-export interface ModuleTypes {
-  [fqn: string]: Interface | Enum | GoClass;
+type ModuleType = Interface | Enum | GoClass;
+type ModuleTypes = ModuleType[];
+
+// TODO: Make this a class?
+export interface Module {
+  root: RootModule;
+  packageName: string;
+  moduleName: string;
+  filePath: string;
+  file: string;
+  submodules: Submodule[];
+  types: ModuleTypes;
+  hasSubmodules(): boolean;
+}
+
+function buildSubmodules(
+  root: RootModule,
+  parent: Module,
+  submodules: readonly JsiiSubmodule[],
+): Submodule[] {
+  return submodules.map((sm) => new Submodule(root, parent, sm));
 }
 
-export class Module {
-  public readonly types: ModuleTypes = {};
-  public readonly submodules: Submodule[];
-  public readonly assembly: Assembly | JsiiSubmodule;
+function unknownType(type: Type): never {
+  throw new Error(`Type: ${type.name} is not an interface, enum, or class`);
+}
 
-  public constructor(assembly: Assembly, submodule?: JsiiSubmodule) {
-    this.assembly = submodule ?? assembly;
-    assembly.types.forEach((type) => {
-      let t: Enum | Interface | GoClass | undefined;
+function buildModuleTypes(parent: Module, types: readonly Type[]): ModuleTypes {
+  return types.map(
+    (type: Type): ModuleType => {
```

Comment:
Nit: Can we use a switch statement on the expression `type.kind` here instead?

Reply:
The problem with a switch on Type.Kind is that typescript can't infer the type of the `Type`.
```
  public get types(): ModuleTypes {
    return this.typeSpec.map(
      (type: Type): ModuleType => {
        switch (type.kind) {
          case TypeKind.Class:
            return new GoClass(this, type);
          // ...
          default:
            throw new Error(
              `Type: ${type.name} with kind ${type.kind} is not a supported type`,
            );
        }
      },
    );
  }
```

Throws a type error `Argument of type 'Type' is not assignable to parameter of type 'ClassType'`. The `isXxxType()` methods on the `jsii-reflect.Type` class are type discriminators so that's why they are used here.

Reply:
Err it's pretty much a discriminated union, I reckon this could be made to work... perhaps the issue is with how we modeled the class hierarchy here that prevents the compiler from making the correct type reduction?

Reply:
we could make it work using `return new GoClass(this, type as Class)`. I generally prefer not to use `as` if possible.